ALSA/ASoC: SOF: Intel: Use PIO mode fro HDA commands instead of CORB on Lunar Lake

ujfalusi opened the pull request:
It is recommended to use Immediate Command Response (PIO mode) instead of CORB on Lunar Lake family for HDA commands.

This PR adds support for PIO mode commands while keeping RIRB functionality for receiving URs

plbossart commented on include/sound/hdaudio.h:
```diff
@@ -390,6 +391,10 @@ int snd_hdac_bus_exec_verb_unlocked(struct hdac_bus *bus, unsigned int addr,
+int snd_hdac_bus_send_cmd_pio(struct hdac_bus *bus, unsigned int val);
+int snd_hdac_bus_get_response_pio(struct hdac_bus *bus, unsigned int addr,
+				  unsigned int *res);
+
```
The commit message is hard to parse due to missing punctuation.

The PIO (Immediate Command Response) can only be used when the CORB is
disabled, inactive but RIRB still can receive unsolicited responses.

Did you mean:
The PIO (Immediate Command Response) can only be used when the CORB is
disabled and hence inactive. RIRB can be enabled independently and can still be used to handle unsolicited responses.

ujfalusi replied:
@plbossart, PIO actually can be used when CORB is enabled but inactive (no commands pending), hence the strange wording.
RIRB on the other hand can be kept enabled. The response will come back to the interface from where the command was sent (IC or CORB).

plbossart commented on sound/hda/hdac_controller.c:
```diff
@@ -135,6 +136,88 @@ static unsigned int azx_command_addr(u32 cmd)
 	return addr;
 }
 
+/* receive a response via Immediate Response */
```
receive an Immediate Response with PIO?

ujfalusi replied:
ok

plbossart commented on sound/soc/sof/intel/hda-bus.c:
```diff
@@ -72,7 +78,14 @@ void sof_hda_bus_init(struct snd_sof_dev *sdev, struct device *dev)
+	} else {
+		snd_hdac_ext_bus_init(bus, dev, &bus_core_pio_ops, sof_hda_ext_ops);
+		bus->use_pio_for_commands = true;
+	}
```
Don't you need a LunarLake-specific quirk for snd-hda-intel as well? CORB/PIO are not SOF-specific.

plbossart replied:
I would actually add a new bit in AZX capabilities and do everything in sound/hda. There's no reason to do anything in SOF regarding commands.

e.g. add something in hda_controller.h 
````
#define  AZX_DCAPS_PIO_COMMANDS 31
````
and add the quirk in the PCI data.

ujfalusi replied:
> Don't you need a LunarLake-specific quirk for snd-hda-intel as well? CORB/PIO are not SOF-specific.

Yes, we need a separate quirk for legacy HDA, or a way to use PIO command in case of Lunar Lake.
This PR is not complete, it is for SOF so we can run complete validation test on it.

ujfalusi replied:
> I would actually add a new bit in AZX capabilities and do everything in sound/hda. There's no reason to do anything in SOF regarding commands.

The two stack is independent in this regards, afaik.
We need separate ways to set different `hdac_bus_ops` via legacy and SOF stack _and_ to set the new `use_pio_for_commands` flag at the same time.

> e.g. add something in hda_controller.h
> 
> ```
> #define  AZX_DCAPS_PIO_COMMANDS 31
> ```

The legacy stack do not support this type of operation (hda_controller w/ azx_*), it can fallback to pure PIO mode w/o RIRB and thus no URs, so the legacy is a bit more involved.

> 
> and add the quirk in the PCI data.

Sure, but the legacy (using hda_controller) and SOF (using hdac_controller) do not share code that we could do this in one place... afaik, imho.

ranj063 commented on sound/hda/hdac_controller.c:
```diff
@@ -135,6 +136,88 @@ static unsigned int azx_command_addr(u32 cmd)
 }
 
+/* receive a response via Immediate Response */
+static int snd_hdac_bus_wait_for_pio_response(struct hdac_bus *bus,
```
@ujfalusi doesnt it make sense to only add the bool in this patch and add the helpers for PIO helpers in the next patch where they're actually used?

ujfalusi replied:
But the bool alone will not do anything, we need the functions which can be used to send the commands via PIO and at the same time to keep the CORB disabled.

ujfalusi committed:
ALSA: hda: Introduce flags to force commands via PIO instead of CORB

Add AZX_DCAPS_PIO_COMMANDS quirk (bit 31) and use_pio_for_commands flag to
be able to select PIO mode as alternative for CORB based command sending
while retaining the RIRB functionality to receive unsolicited responses.

This mode differs from the azx single_cmd mode when RIRB is disabled.

The mixed mode is needed on Lunar Lake family because it is recommended to
use Immediate Command Response (PIO mode) instead of CORB for HDA commands.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ALSA: hda: hdac_controller: Implement support for use_pio_for_commands mode

In case the use_pio_for_commands flag is set we must not enable the
CORB DMA to make sure that it is not interfering with the immediate
command mode.

Convert the snd_hdac_bus_send_cmd/snd_hdac_bus_get_response as wrappers to
call either the PIO or CORB based command handling depending on the
use_pio_for_commands flag.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ALSA: hda: Intel: Select AZX_DCAPS_PIO_COMMANDS for Lunar Lake

It is recommended that on Lunar Lake the PIO (immediate command response)
is used instead of CORB/RIRB for commands/verbs.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi committed:
ASoC: SOF: Intel: hda-bus: Use PIO mode for Lunar Lake

It is recommended that on Lunar Lake the PIO (immediate command response)
is used instead of CORB/RIRB for commands/verbs.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v1:
- separate the flag introduction
- Add quirk flag for legacy to use: AZX_DCAPS_PIO_COMMANDS
- do not export the new PIO send_cmd/get_response and select between CORB/RIRB and PIO in core code based on use_pio_for_commands flag
- enable the PIO mode for LNL in legacy and SOF stack.
- dropped the SOF patch to use snd_hdac_bus_exec_verb_unlocked() as the PIO vs CORB selection is done in core now.

ujfalusi commented:
Changes since v2:
- fix prototype comment for snd_hdac_bus_get_response_rirb()

plbossart commented on sound/pci/hda/hda_intel.c:
```diff
@@ -2503,7 +2503,8 @@ static const struct pci_device_id azx_ids[] = {
 	/* Lunarlake-P */
-	{ PCI_DEVICE_DATA(INTEL, HDA_LNL_P, AZX_DRIVER_SKL | AZX_DCAPS_INTEL_SKYLAKE) },
+	{ PCI_DEVICE_DATA(INTEL, HDA_LNL_P, AZX_DRIVER_SKL | AZX_DCAPS_INTEL_SKYLAKE |
+					    AZX_DCAPS_PIO_COMMANDS) },
```
we could also introduce a macro such as  AZX_DRIVER_LNL where all the defaults are set, that will save lines for future additions.

ujfalusi replied:
Right, my optimistic thinking was that this is only going to be needed for LNL, but ACE2.0 variants will need this as well.
OK.

plbossart commented on sound/soc/sof/intel/hda-bus.c:
```diff
@@ -72,7 +72,12 @@ void sof_hda_bus_init(struct snd_sof_dev *sdev, struct device *dev)
+
 	snd_hdac_ext_bus_init(bus, dev, &bus_core_ops, sof_hda_ext_ops);
+
+	if (chip && chip->hw_ip_version == SOF_INTEL_ACE_2_0)
```
"> SOF_INTEL_ACE_1_0" would be more future-proof

ujfalusi replied:
I hope that  `> ACE_2_0` will not need this, if it does then we change the condition.

plbossart commented on sound/pci/hda/hda_controller.c:
```diff
@@ -1180,6 +1180,9 @@ int azx_bus_init(struct azx *chip, const char *model)
 
+	if (chip->driver_caps & AZX_DCAPS_PIO_COMMANDS)
+		bus->core.use_pio_for_commands = true;
+
```
I would split the library from the sound/pci/hda/hda_controller.c in two separate patches.

Arguably the sound/pci/hda/hda_controller.c stuff should be combined with the next patch, it would make it easier to bisect regressions or changes in behavior (firmware download, etc).

How many engineers does it take to split code, eh?

ujfalusi replied:
The problem is that the only place where the CORB DMA is enabled is via the `snd_hdac_bus_init_cmd_io()` in hdac_controller.c and that needs to be done at the same time as we allow hda_controller to use PIO commands.

Hrm, unless I split the patch and have the hdac_controller.c changes in the first one which anyways also needs this.

ujfalusi replied:
I will still keep the enabling this as separate patches.

plbossart reviewed (commented):
I only have cosmetic comments @ujfalusi this looks pretty good to me!

ujfalusi committed:
ALSA: pci: hda: hda_controller: Add support for use_pio_for_commands mode

Set the use_pio_for_commands flag in case AZX_DCAPS_PIO_COMMANDS quirk is
enabled.

When the PIO command mode is used we can re-use the existing
azx_single_send_cmd() / azx_single_get_response() functions safely as the
CORB DMA is not going to be enabled in snd_hdac_bus_init_cmd_io().

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v3:
- split the hdac_controller/hda_controller patch which implements support for use_pio_for_commands
- Create AZX_DCAPS_INTEL_LNL for the quirk definition
- taken out from draft

plbossart commented:
@ujfalusi should we merge this?

ujfalusi commented:
> @ujfalusi should we merge this?

I think we should and send it for 6.10 as early as possible to get feedback on the legacy HDA change.

marc-hb commented:
Revert just submitted in:
- #4892

... to avoid these mass test failures:
- #4889
